docs/langref/relay_expr.rst

```diff
@@ -0,0 +1,402 @@
+types. In particular, operator types are registered using a single
+type relation (see the above subsection), typically a relation
+specialized to that operator. For example, the :code:`add` operator
+is registered with the `Broadcast` relation, indicating that the
+arguments of :code:`add` must be tensors and that the return type
+is a tensor whose shape depends on those of its arguments.
+
+Operators are rendered without a sigil (e.g :code:`add`, :code:`subtract`)
+when pretty-printing Relay programs.
+Operators are explicitly contained in the program and are uniquely
+identifiable by pointer during a run of the Relay compiler.
+
+See :py:class:`~tvm.relay.op.Op` for the definition and documentation
+of operator nodes, demonstrating the infrastructure for registering
+operator metadata. The other files in :py:class:`~tvm.relay.op` give
+handles for generating a call to various pre-registered operators.
+
+Dataflow Fragment
+=================
+
+This subsection covers the set of Relay expressions that do not involve
+control flow. That is, any portion of a program comprised only of these
+expressions corresponds to a pure computation graph without control flow.
+Note that global and local variables are also part of the dataflow fragment.
+
+Constants
+~~~~~~~~~
+
+This node represents a constant tensor value
+(see :py:mod:~tvm.relay.Value for more details).
+The constants are represented as :py:class:`~tvm.NDArray`,
+allowing Relay to utilize TVM operators for constant evaluation.
+
+See :py:class:`~tvm.relay.expr.Constant` for its definition and documentation.
+
+Tuple
+~~~~~
+
+The tuple node builds a finite (that is, of statically known size) sequence of heterogeneous data. 
+These tuples match Python's closely. Their fixed length allows for efficient projection of their
+members.
+
+.. code-block:: python
+
+    (a, b, c) : Tuple<A, B, C>
+
+    (add(add(a, b), c), d) : Tuple<Tensor<float32, (10, 10)>, Tensor<float32, (100, 100)>>
+
+See :py:class:`~tvm.relay.expr.Tuple` for its definition and documentation.
+
+Tuple Projection
+~~~~~~~~~~~~~~~~
+
+A tuple must be indexed by an integer constant in order to extract a
+particular member of the tuple. Projections are 0-indexed.
+
+For example, the below projection evaluates to :code:`b`:
+
+.. code-block:: python
+
+   (a, b, c).1
+
+See :py:class:`~tvm.relay.expr.TupleGetItem` for its definition and documentation.
+
+Function Expressions
+~~~~~~~~~~~~~~~~~~~~
+
+Functions are first class in Relay and can be used in any expression
+position. Function expressions are the same as global functions and
+use nearly the same syntax (besides not including a name), but do
+not have a globally unique name.
+
+.. code-block:: python
+
+    fun (%x : Tensor<float32, (10, 10)>, y: Tensor<float32, (10, 10)>
+                -> Tensor<float32, (10, 10)> { add(%x, %y) }
+
+Note that function expressions evaluate to closure values. Closures
+store values for all free variables in their body. A free variable
```

Comment:
Closures are values which are represented as a pair of a local environment, containing the variables defined outside the scope of the function's body and the function itself.

Reply:
Rewritten in e7c5a04